KAFKA-9487: Follow-up PR of Kafka-9445

brary opened the pull request:
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

brary committed:
Addressing code review comments of 9445

guozhangwang reviewed (commented):
The non unit-testing part LGTM. @mjsax could you take a look at the unit tests and see if they have addressed all your comments?

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java:
```diff
@@ -21,12 +21,11 @@
 
 /**
- * Represents all the query options that a user can provide to state what kind of stores it is expecting.
+ * Represents all the query options that a user can provide to state what kind of stores we are expecting.
```
cosmetic: `Represents all the query options that a user can provide to control the kind of stores we would like to fetch` ?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java:
```diff
@@ -21,12 +21,11 @@
 /**
- * Represents all the query options that a user can provide to state what kind of stores it is expecting.
+ * Represents all the query options that a user can provide to state what kind of stores we are expecting.
  * The options would be whether a user would want to enable/disable stale stores
```
Cosmetic: ` Options could indicate - whether we want stale stores to be included (or) we want to filter for only stores belonging to a partition`.

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java:
```diff
@@ -35,46 +34,42 @@
     public static <T> StoreQueryParams<T> fromNameAndType(final String storeName,
                                                           final QueryableStoreType<T>  queryableStoreType) {
-        return new StoreQueryParams<T>(storeName, queryableStoreType);
+        return new StoreQueryParams<T>(storeName, queryableStoreType, null, false);
```
Could we use a different sentinel than `null` for the `partition` field?. Does an `Optional` make sense?

brary replied:
Actually we had a long discussion on keeping this null and adding it in Java doc that null means all partitions. Do you think this is important to start a discussion on the thread again? :)

vinothchandar replied:
dont worry about it :) .. Happy to budge on stylistic things and leave it to the committers' call

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java:
```diff
@@ -35,46 +34,42 @@
      * @param partition   The specific integer partition to be fetched from the stores list by using {@link StoreQueryParams}.
      *
-     * @return String storeName
+     @return StoreQueryParams a new {@code StoreQueryParams} instance configured with the specified partition
```
nit: we need the `*` before `@return` right?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1148,6 +1148,17 @@ private static Pattern buildPattern(final Collection<String> sourceTopics,
+                return results;
+            }
+        }
+        return null;
```
any reason we can't use an empty list for the sentinel?

brary replied:
Done.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1148,6 +1148,17 @@ private static Pattern buildPattern(final Collection<String> sourceTopics,
+    public List<String> sourceTopicsForStore(final String storeName) {
+        final List<String> results = new ArrayList<>();
+        for (final Map.Entry<String, Set<String>> entry : stateStoreNameToSourceTopics.entrySet()) {
+            if (entry.getKey().equals(storeName)) {
```
can't we just lookup the map? why iterate over if we are going to compare the key anyway?

brary replied:
I think the return type of stateStoreNameToSourceTopics is a set and we wanted to return a list, so I had to loop and add it to list.

vvcephei replied:
I'm also a little confused. Would this not be the same?
```
results.addAll(maybeDecorateInternalSourceTopics(stateStoreNameToSourceTopics.get(storeName))
```
I.e., converting the returned set to a list doesn't seem to require iterating over the whole map just to look for one key.

On a different note, it doesn't seem like we really need a List in particular. We could just relax it to Collection, and then you wouldn't have to convert the set to a list. But that's not a big issue to me.

brary replied:
Makes sense, Changed to Collection and removed the loop.

vinothchandar commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1148,6 +1148,17 @@ private static Pattern buildPattern(final Collection<String> sourceTopics,
         return results;
     }
 
+    public List<String> sourceTopicsForStore(final String storeName) {
```
a side note: Should we also consider the map `stateStoreNameToSourceRegex` here?

brary replied:
I am not sure, do you think it will identify more topic groups? At the end of the day we are identifying a topic groupId to which a store belongs to. 

On a side note, if we do want to add stateStoreNameToSourceRegex along with stateStoreNameToSourceTopics, then we can probably return a set instead of list in this function and iterate over both of above one by one.

vinothchandar reviewed (commented):
Left few comments..

brary committed:
Addressing code review comments

brary commented:
> Left few comments..

Thanks @guozhangwang and @vinothchandar . Addressed a few comments and have replied to some left.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/KafkaStreams.java:
```diff
@@ -1171,11 +1171,9 @@ public void cleanUp() {
      * The returned object can be used to query the {@link StateStore} instances.
      *
-     * @param storeQueryParams   to set the optional parameters to fetch type of stores user wants to fetch when a key is queried
+     * @param storeQueryParams   the parameteres used to fetch a type of queryable store
```
```suggestion
     * @param storeQueryParams   the parameters used to fetch a queryable store
```

brary replied:
Added this my commit for addressing the other comments.

vvcephei commented on streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java:
```diff
@@ -21,12 +21,11 @@
+ * The options could indicate - whether we want stale stores to be included (or)
+ * we want to filter for only stores belonging to a partition.
+ * If no specific partition is specified the default behavior is to fetch the stores for all the partitions
  * available on that instance for that particular store name.
```
This comment can easily become outdated as the class evolved. Better to keep it high level and defer to the setters' javadocs for a description of the available operations.

Eg:
`{@code StoreQueryParams} allows you to pass a variety of parameters when fetching a store for interactive query.`

brary replied:
Done.

vvcephei reviewed (changes requested):
Hey, thanks for the update @brary .

IIUC, the main motivation for this PR was some follow-up comments from @mjsax . It would be good to get a +1 from him before merging. Maybe after the other comments are addressed.

brary committed:
Addressing code review comments

brary commented:
> Hey, thanks for the update @brary .
> 
> IIUC, the main motivation for this PR was some follow-up comments from @mjsax . It would be good to get a +1 from him before merging. Maybe after the other comments are addressed.

Thanks @vvcephei for the comments. I have made the changes suggested by you as well. 

@mjsax  I guess everyone else has reviewed and their comments have been incorporated. It would be great if you can give one last look and if it's okay merge it.

brary committed:
Merge branch 'trunk' into navinderbrar-Followup-KAFKA-9445

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java:
```diff
@@ -35,46 +30,42 @@
-        storeQueryParams.partition = partition;
-        storeQueryParams.staleStores = this.staleStores;
-        return storeQueryParams;
+        return new StoreQueryParams<T>(this.storeName(), this.queryableStoreType(), partition, this.staleStores);
```
Nit: No need to do method calls, we can just access the variables directly.  Also, try to avoid unnecessary `this.` prefix

brary replied:
Done.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/StoreQueryParams.java:
```diff
@@ -35,46 +30,42 @@
-        storeQueryParams.partition = this.partition;
-        storeQueryParams.staleStores = true;
-        return storeQueryParams;
+        return new StoreQueryParams<T>(this.storeName(), this.queryableStoreType(), this.partition, true);
```
as above.

brary replied:
Done.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/state/internals/StreamThreadStateStoreProvider.java:
```diff
@@ -34,6 +34,8 @@
 import java.util.Map;
 import java.util.Set;
+import java.util.Collection;
+import java.util.stream.Collectors;
```
I cannot comment below, but I think we should still simplify the code when the user specifies a partition. I left a comment on the original PR about it (https://github.com/apache/kafka/pull/7984/files#r375513526). Let me know if this comment makes sense.

brary replied:
Thanks. It makes sense, I have made the suggested change.

mjsax reviewed (commented):
Thanks for the PR. Couple of minor comments. Also added one more reply on the original PR that should be addressed in this PR (https://github.com/apache/kafka/pull/7984/files#r375517935).

vvcephei reviewed (commented):
Thanks for addressing my comments!

mjsax commented:
@vinothchandar @vvcephei I would still propose to rename the new class to `StoreQueryParameters` -- it's not a big deal to rename even after feature freeze IHMO.

Thoughts?

vinothchandar commented:
`StoreQueryParameters` does sound better to me. but leave it you all. :)

brary committed:
Changing StoreQueryParams to StoreQueryParameters

brary committed:
Resolving merge conflicts

brary commented:
@mjsax  I have changed the class to StoreQueryParameters and addressed the comment you gave in earlier PR. There were few checkstyles which were failing in ValueAndTimestampSerializerTest unrelated to this PR. I have fixed those as well.

mjsax reviewed (approved):
Thanks @brary! LGTM. I leave it to @vvcephei to merge this PR to `trunk` and cherry-pick to `2.5` branch.

vvcephei commented:
Thanks @mjsax . I'll take care of it tomorrow.

I don't see any test results, so I'll ask Jenkins to re-run it:

Retest this, please.

vvcephei commented:
Unrelated test failures:
kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdAllGroups
kafka.admin.DeleteConsumerGroupsTest.testDeleteCmdWithMixOfSuccessAndError

Proceeding to merge.

vvcephei commented:
Cherry-picked onto 2.5 as 3e77458cfb0f7a7586f17777f0e775d91e4744e1
